### PR TITLE
TF-3789: Cleanup + Launch_manager support

### PR DESCRIPTION
**Pull request by PGlaubitzWork**
First set of changes is "Reduce duplicated logic, compile with C++11", so that support for initial config should appear in one location rather than several.

### CODE REVIEW

File: robotiq_force_torque_sensor/CMakeLists.txt [hunk 1]
```diff
@@ -1,6 +1,12 @@
 cmake_minimum_required(VERSION 2.8.3)
 project(robotiq_force_torque_sensor)
 
+# "-Werror" removed since the code originates as thrid party and can't compile with it added
```
Comment (adamcalhoon) [inline robotiq_force_torque_sensor/CMakeLists.txt]:
```suggestion
# "-Werror" removed since the code originates as third party and can't compile with it added
```

File: robotiq_force_torque_sensor/nodes/rq_sensor.cpp [hunk 2]
```diff
@@ -171,206 +288,41 @@ static robotiq_force_torque_sensor::ft_sensor get_data(void)
...
-    if (connected) {
-        connected = try_connection();
-    }
+    std::unique_ptr<SensorBase> sensor_ptr;
 
-    //Starts the stream
-    if (connected) {
-        connected = try_connection();
+    bool simulated;
+    ros::param::param<bool>("/simulate_robot", simulated, false);
```
Comment (adamcalhoon) [inline robotiq_force_torque_sensor/nodes/rq_sensor.cpp]: There is a `ros::param::param<T>` that takes 2 args -- path, default value and returns the result. I find that more readable.
```suggestion
    bool simulated = ros::param::param<bool>("/simulate_robot", false);
```

File: robotiq_force_torque_sensor/nodes/rq_sensor.cpp [hunk 3]
```diff
@@ -40,99 +41,224 @@
...
+    bool _getSensorData(robotiq_force_torque_sensor::ft_sensor &msgStream) override
+    {
+        msgStream.Fx = 0.0;
+        msgStream.Fy = 0.0;
+        msgStream.Fz = 0.0;
+        msgStream.Mx = 0.0;
+        msgStream.My = 0.0;
+        msgStream.Mz = 0.0;
+        // Keep publishing rate to 100Hz
+        ros::Duration(0.01).sleep();
```
Comment (adamcalhoon) [inline robotiq_force_torque_sensor/nodes/rq_sensor.cpp]: ros::Rate will adjust its sleep time to try to make the rate actually 100Hz rather than 100+Hz you'd see with ros::Duration

```suggestion
        ros::Rate(100).sleep();
```

File: robotiq_force_torque_sensor/nodes/rq_sensor.cpp [hunk 4]
```diff
@@ -40,99 +41,224 @@
...
+        return true;
+    }
+    void _decodeMessageAndDo(INT_8 const *const /*buff*/, INT_8 *const /*ret*/) override
```
Comment (adamcalhoon) [inline robotiq_force_torque_sensor/nodes/rq_sensor.cpp:141]:
```suggestion

    void _decodeMessageAndDo(INT_8 const *const /*buff*/, INT_8 *const /*ret*/) override
```

File: robotiq_force_torque_sensor/CMakeLists.txt (hunk 1 above)
Comment (adamcalhoon) [reply robotiq_force_torque_sensor/CMakeLists.txt]: Also :frowning_face:

**Review by adamcalhoon — APPROVED** [review]
A couple little gripes:
- The first one I mentioned in person: I wish we could just use the NullForceSensor implementation in the `simulate` case, but I'm OK kicking that can down the road.
- I think it would be easier for my brain if the SensorBase implementation came before the other sensor implementations, but perhaps that's personal preference.

File: robotiq_force_torque_sensor/CMakeLists.txt (hunk 1 above)
Comment (PGlaubitzWork) [reply robotiq_force_torque_sensor/CMakeLists.txt]: :man_facepalming:

File: robotiq_force_torque_sensor/nodes/rq_sensor.cpp (hunk 3 above)
Comment (adamcalhoon) [reply robotiq_force_torque_sensor/nodes/rq_sensor.cpp]: D'oh! Temporaries won't work!

Comment (PGlaubitzWork) [reply robotiq_force_torque_sensor/nodes/rq_sensor.cpp]: (D'oh)^2 "100+Hz you'd see with ros::Duration" -> Since time is in the denominator it'll be less than 100Hz (like ~98) :stuck_out_tongue_winking_eye:

**Review by adamrr — APPROVED** [review]

File: robotiq_force_torque_sensor/CMakeLists.txt [hunk 5]
```diff
@@ -34,7 +41,10 @@ generate_messages(
 ## catkin specific configuration ##
 ###################################
 catkin_package(
-  CATKIN_DEPENDS message_runtime std_msgs
+  CATKIN_DEPENDS
+    message_runtime
+    std_msgs
+    launch_manager_interface
```
Comment (mechterp10) [inline robotiq_force_torque_sensor/CMakeLists.txt:47]:
```suggestion
```
I believe you should only require this if dependent projects need this package. In that case, those project should have an explicit dependency themselves

File: robotiq_force_torque_sensor/nodes/rq_sensor.cpp [hunk 6]
```diff
@@ -40,99 +41,233 @@
...
+public:
+    SensorBase() : _connection_pub(_nh.advertise<std_msgs::Bool>("robotiq_force_torque_sensor_connected", 1))
+    {
+        ros::param::param<std::string>("~frame_id", _wrench_msg.header.frame_id, "robotiq_force_torque_frame_id");
+    }
+
+    virtual ~SensorBase() = 0;
+    virtual bool makeConnection() = 0;
+
+    void Connect();
```
Comment (mechterp10) [inline robotiq_force_torque_sensor/nodes/rq_sensor.cpp:69]:
```suggestion
    void connect();
```

File: robotiq_force_torque_sensor/nodes/rq_sensor.cpp [hunk 7]
```diff
@@ -40,99 +41,233 @@
...
+    while (ros::ok() && tries_left > 0) {
+        if (tries_left-- % 2 == 0) {
             ROS_INFO("Waiting for sensor connection...");
         }
-        usleep(10000); //Attend 1 seconde.
-
-        ret = rq_sensor_state(max_retries_);
-        if(ret == 0)
-        {
+        if (rq_sensor_state(_max_retries) == 0) {
```
Comment (mechterp10) [inline robotiq_force_torque_sensor/nodes/rq_sensor.cpp:266]: You removed the 0.01s sleep here. For now, I would leave it in

File: robotiq_force_torque_sensor/nodes/rq_sensor.cpp [hunk 8]
```diff
@@ -40,99 +41,233 @@
...
             ROS_INFO("Sensor connected!");
             return true;
         }
-        count++;
-        ros::spinOnce();
+        ros::Duration(1.0).sleep();
```
Comment (mechterp10) [inline robotiq_force_torque_sensor/nodes/rq_sensor.cpp:270]: Why did you add this 1s sleep here?

File: robotiq_force_torque_sensor/nodes/rq_sensor.cpp [hunk 9]
```diff
@@ -171,206 +297,49 @@ static robotiq_force_torque_sensor::ft_sensor get_data(void)
...
 
-    // Publish the connection state
-    connection_msg.data = connected;
-    connection_pub.publish(connection_msg);
-    ros::spinOnce();
-
-    if (connected)
-    {
-        ROS_INFO("Starting Sensor");
+    sensor_ptr->Connect();
```
Comment (mechterp10) [inline robotiq_force_torque_sensor/nodes/rq_sensor.cpp:324]:
```suggestion
    sensor_ptr->connect();
```

File: robotiq_force_torque_sensor/nodes/rq_sensor.cpp [hunk 10]
```diff
@@ -171,206 +297,49 @@ static robotiq_force_torque_sensor::ft_sensor get_data(void)
...
 
-    if (!connected)
-    {
-        connection_msg.data = connected;
-        connection_pub.publish(connection_msg);
-        ros::spinOnce();
-    }
-
-    ROS_INFO("Sensor Stopped");
-    if (created_publishers)
```
Comment (mechterp10) [inline robotiq_force_torque_sensor/nodes/rq_sensor.cpp:347]: Please leave this in unless testing proves that it is not required when undergoing rapid node restarts.

**Review by mechterp10 — CHANGES_REQUESTED** [review]
I am drastically concerned about the larger refactor in place here. A lot of the checks and timing which were kept in place were largely following legacy items from the original implementation from Robotiq. I am uncomfortable with this refactor being merged in an opposed to the smaller work for initial config without a large amount of testing. Additionally, this node can be arbitrarily restarted by ready_attachment_manager when it dies, (e.g. someone unplugs it) and I am not sure how that interaction currently works with the refactor.

File: robotiq_force_torque_sensor/nodes/rq_sensor.cpp (hunk 8 above)
Comment (PGlaubitzWork) [reply robotiq_force_torque_sensor/nodes/rq_sensor.cpp:270]: This is the moved sleep value set to match comment of moved line.

File: robotiq_force_torque_sensor/nodes/rq_sensor.cpp (hunk 7 above)
Comment (PGlaubitzWork) [reply robotiq_force_torque_sensor/nodes/rq_sensor.cpp:266]: Sleep was moved such that it would be done only if the first attempt failed. Rather than before every attempt, if first attempt is likely to fail w/o it previous order can be restored.

File: robotiq_force_torque_sensor/nodes/rq_sensor.cpp (hunk 10 above)
Comment (PGlaubitzWork) [reply robotiq_force_torque_sensor/nodes/rq_sensor.cpp:347]: Implementation of  publishers issue shutdown on destruction, so should be redundant.

**Comment by PGlaubitzWork** [comment]
> I am drastically concerned about the larger refactor in place here...
This is one of the reasons I asked you if there was a reason to not refactor and clean up before I started.
